### PR TITLE
Add FLOAT4E2M1 data type

### PR DESCRIPTION
**Pull request by yuanyao-nv**
### Description
- Add FLOAT4E2M1 as a new data type to proto as well as relevant helper functions and tests.
- This PR splits out the portion of https://github.com/onnx/onnx/pull/6283 relevant to data type updates to reduce the PR's size.

### Motivation and Context
Narrow precision data types with sub-byte bit widths are becoming solutions to the rising cost, performance, and deployment challenges of LLMs. ONNX already has INT4/UINT4. FP4 is another commonly used narrow-precision data type for compressing both the weights and activations of LLMs. For example [OCP](https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf) MXFP4 uses E2M1 as element type.

Similar to INT4/UNIT4, FP4 weights/inputs are expected to be packed.

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]

File: onnx/backend/test/data/node/test_reduce_log_sum_exp_keepdims_random/test_data_set_0/output_0.pb
Comment (justinchuby) [inline onnx/backend/test/data/node/test_reduce_log_sum_exp_keepdims_random/test_data_set_0/output_0.pb:1]: @yuanyao-nv I just noticed - the test data shouldn’t change in this PR?

Comment (yuanyao-nv) [reply onnx/backend/test/data/node/test_reduce_log_sum_exp_keepdims_random/test_data_set_0/output_0.pb:1]: This was auto generated by `update_doc.sh`. I've seen cases where some seemingly irrelevant pb files get updated by  the script. Do you understand why?

**Comment by lutzroeder** [comment]
@uanyao-nv can please you share a sample `.onnx` file using the `float4e2m1` data type.

File: onnx/backend/test/data/node/test_reduce_log_sum_exp_keepdims_random/test_data_set_0/output_0.pb
Comment (justinchuby) [reply onnx/backend/test/data/node/test_reduce_log_sum_exp_keepdims_random/test_data_set_0/output_0.pb:1]: The pb files can be different in bytes when generated on different OS. I think that’s what’s happening here.

Comment (yuanyao-nv) [reply onnx/backend/test/data/node/test_reduce_log_sum_exp_keepdims_random/test_data_set_0/output_0.pb:1]: I see. I wonder if it's possible to let the pipeline auto-generate these test files instead and reduce the number of files changed in each PR.

**Pull request merged**